volatility3/framework/plugins/windows/mftscan.py

```diff
@@ -174,126 +190,120 @@ def run(self):
         )
 
 
-class ADS(interfaces.plugins.PluginInterface):
+class ADS(MFTScan):
     """Scans for Alternate Data Stream"""
 
     _required_framework_version = (2, 0, 0)
 
-    @classmethod
-    def get_requirements(cls):
-        return [
-            requirements.TranslationLayerRequirement(
-                name="primary",
-                description="Memory layer for the kernel",
-                architectures=["Intel32", "Intel64"],
-            ),
-            requirements.VersionRequirement(
-                name="yarascanner", component=yarascan.YaraScanner, version=(2, 0, 0)
-            ),
-        ]
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # which DATA attribute should be displayed
+        self._display_first_data = False
+
+    def _parse_data_record(self, mft_record, attr):
+        if attr.Attr_Header.NonResidentFlag:
+            return
+
+        # regular $DATA
+        if self._display_first_data:
+            content = attr.get_resident_filecontent()
+            if content:
+                content = format_hints.HexBytes(content)
+            else:
+                content = renderers.NotAvailableValue()
+
+            yield 0, (
+                format_hints.Hex(self._record_map[mft_record.RecordNumber][2]),
+                mft_record.get_signature(),
+                mft_record.RecordNumber,
+                attr.Attr_Header.AttrType.lookup(),
+                self._record_map[mft_record.RecordNumber][0],
+                content,
+            )
+
+        # ADS $DATA
+        elif attr.Attr_Header.NameLength > 0:
+            ads_name = attr.get_resident_filename()
+            if not ads_name:
+                ads_name = renderers.NotAvailableValue()
+
+            content = attr.get_resident_filecontent()
+            if content:
+                content = format_hints.HexBytes(content)
+            else:
+                content = renderers.NotAvailableValue()
+
+            yield 0, (
+                format_hints.Hex(self._record_map[mft_record.RecordNumber][2]),
+                mft_record.get_signature(),
+                mft_record.RecordNumber,
+                attr.Attr_Header.AttrType.lookup(),
+                self._record_map[mft_record.RecordNumber][0],
+                ads_name,
+                content,
+            )
+
+    def parse_data_records(self, mft_record, attr):
```

Comment:
This would also be a generator, even though you can't tell until the very end of the function.

Reply:
Fixed